neps/nep-0413.md

```diff
@@ -0,0 +1,135 @@
+
+### Why Off Chain?
+So the user would not incur in GAS fees, nor the signed authentication gets broadcasted into a public network.
+
+### Why The Message Being Signed MUST NOT be a Transaction?
+An attacker could make the user inadvertently sign a valid transaction which, once signed, could be submitted into the network to execute it.
+
+### Why The Message Needs to Include a Service Identifier?
+To avoid replay attacks, i.e. a malicious app requesting the user to authenticate in their service, only to use the authentication message in a third-app. 
+
+Including the service's identifier (e.g. domain name) and making sure the user knows about it should mitigate this kind of attacks.
+
+### Why a FullAccess Key? Why Not Simply Creating an [FunctionCall Key](https://docs.near.org/concepts/basics/accounts/access-keys) for Signing?
+The most common flow for [NEAR user authentication into a Web3 frontend](https://docs.near.org/develop/integrate/frontend#user-sign-in--sign-out) involves the creation of a [FunctionCall Key](](https://docs.near.org/concepts/basics/accounts/access-keys)).
+
+One might feel tempted to reproduce such process here, for example, by creating a key that can only be used to call a non-existing method in the user's account. This is a bad idea for several reasons:
+1. It implies that the user needs to expend gas in creating a new key.
+2. Since any third-party can ask the user to create a `FunctionCall Key`, this opens a vector of attack.
+
+Using a FullAccess key allows us to be sure that the message was signed by the user (since nobody should have access to their `FullAccess Key`), while keeping the constraints of not expending gas in the process (because no new key needs to be created).
+
+### How to Return the Signed Message in a Safe Way
+Sending the signature in a query string to an arbitrary URL (even within the correct domain) is not secure as the data can be leaked (e.g. through headers, etc). Using URL fragments instead will improve security, since the fragment isn't included in the Referrer.
+
+### NEAR Signatures
+NEAR transaction signatures are not plain Ed25519 signatures but Ed25519 signatures of a SHA-256 hash (see [near/nearcore#2835](https://github.com/near/nearcore/issues/2835)). Any protocol that signs anything with NEAR account keys should use the same signature format.
+
+## Specification
+Given the previous [rationales](#rationale-and-alternatives), this NEP proposes the wallets to implement a `verifyOwner` method following the specifications bellow.
+
+### Goal
+The `verifyOwner` method must take as input a `domain` and a `message`, and transform them into a signature.
+
+### Input Interface
+In order to sign a message, `verifyOwner` must implement the following input interface:
+
+```jsx
+interface VerifyOwnerParams {
+  domain: string; // The app in which the user wants to authenticate, e.g. myapp.com.
+  message: string; // A message to be included in the signature
```

Comment:
```suggestion
  nonce: [u8; 256] ; // A unique random nonce to be included in the signature
```

Reply:
I would say `A nonce to be included in the signature`, since the wallet has no way to know if it is random or unique.

Reply:
Why is the `message` being removed in favor of `nonce`? This greatly narrows the usecases the method can be used for. While this is often desired thing, not sure that the community will agree this time.

Having custom application defined `message` parameter allows the application to use the signature of the message as proof for ownership also of what is inside the contents of the `message`. 

### Example

The example application is "Web3 Messenger" and it is end-to-end encrypted messaging application for Web3 identities.

Alice wants to register to this application with her Web3 identity of `alice.near` so she can send and receive messages.

The application requires from the Web3 identities to provide `encryptionKey` that will be used to encrypt the messages that are send to them and proof of ownership that link the identity and the key.

Alice as owner of `alice.near`, is generating new messaging keypair consisting of `encryptionKey` (PublicKey) and `decryptionKey` (PrivateKey).

To register her `alice.near` Web3 identity and the corresponding `encryptionKey` to the application, the application requires from Alice to `verifyOwner` for `alice.near` with custom message of

```
"Me 'alice.near' want to receive messages via 'Web3 Messenger' encrypted with '0x123456789'"
```

Once this is signed, the application has verified that Alice owns the `alice.near` identity and also has approved the usage of the `0x123456789` as `encryptionKey` for this identity. This signature (proof) and `encryptionKey` is stored in the applications storage (registry).

When `bob.near` wants to send message to `alice.near`, lookup is made in the application to get `alice.near`'s `encryptionKey`, once found it is used to encrypt the sent message. Once `alice.near` receives the message, the secret `decryptionKey` is used to decrypt the message. 

If `bob.near` doesn't trust that the key is Alice's, he can reconstruct the whole signature payload and use the `alice.near`'s `publicKey` to verify the signature matches.

**Sorry for the long and maybe confusing example**, but simply put, having the ability to put custom message allows to verify the ownership of additional resources, not just the account ownership.

Reply:
@toteto sending messages between the server and the user is out of the scope of this NEP. We want to `authenticate the user` and that's it.

In the case of the server and user wanted to talk, they can do it after the user has been authenticated through a secure channel that goes out of the scope of this NEP.

Reply:
Hi @gagdiez thanks for drafting this proposal. I think there is some confusion with the `message` field name and the use cases of `messaging` @toteto  described. The core use case of signing `message`/`data` is for applications to later prove that the account owner signed it. This lets application developers build important features on top near accounts.

Only letting applications sign a `nonce` associated with a `domain` limits the use cases of signatures. Meaning issues like https://github.com/near/wallet-selector/issues/405 are not resolved as there is not `data` field.  Its likely developers will need to work around this by hashing their `data` and putting the hash in the `nonce`. This would be to the detriment of user experience and safety as they user would need to trust that the application hashed the value correctly.

I Suggest adding `data: string` or `data: Map<string, string> `field to the payload. We can also keep the `nonce` field. The `nonce` can be signed in isolation or used as an extra check to increment, expire, or add uniqueness to the `data` field.

_note: I discussed this response with @toteto before posting._